### ISSUE TITLE
Consider leaving root env on PATH with activate

### ISSUE THREAD

**Issue by kalefranz** [conda/conda#2502]
Advantages:
- This would be more consistent with unix behavior of how PATH is built up during login with append/prepend exports.  Consider a non-python package--e.g. git.  If I install `conda install git` in my root env because I know conda's git is more up-to-date than my system git, and then activate a conda env, I definitely don't want to fall back to the system `/usr/bin/git` while working in a non-root conda env.
- It's more consistent with virtualenv/rbenv/nodenv behavior.
- It's a first step in the direction of "stacked" envs.

Implications:
- Potential confusion with something like ipython (basically different versions of python) was mentioned as a reason for removing the root env from PATH on activate.  However, with activate, we don't completely overwrite PATH, so users could still encounter confusion with other pythons, say system /usr/bin/python or homebrew /usr/local/bin/pip on OS X.
- symlinking conda/activate/deactivate wouldn't be necessary anymore.  Keeping the root env on PATH would mean that conda and activate are already naturally available for use within a conda env.

CC @pzwang let's discuss in person sometime soon

**Comment by janschulz** [comment]
Just for reference: this already happens on Windows:

```
C:\Users\jschulz
λ path # default as coming from a miniconda install
PATH=[...] C:\portabel\miniconda;C:\portabel\miniconda\Scripts;C:\portabel\miniconda\Library\bin;[...]

C:\Users\jschulz
λ activate borked35
Deactivating environment "C:\portabel\miniconda"...
Activating environment "C:\portabel\miniconda\envs\borked35"...

C:\Users\jschulz
[borked35] λ path
PATH=C:\portabel\miniconda\envs\borked35;C:\portabel\miniconda\envs\borked35\Scripts;C:\portabel\miniconda\envs\borked35\Library\bin;[...]C:\portabel\miniconda\Scripts
```

I don't know where the `git` conda package installs itself, but the `miktex` package installs into `Scripts` so that is available and so are the jupyter commands:

```
C:\Users\jschulz
λ ls C:\portabel\miniconda\Scripts |grep jupyter |grep exe
jupyter-console.exe*
jupyter-kernelspec.exe*
jupyter-migrate.exe*
jupyter-nbconvert.exe*
jupyter-qtconsole.exe*
jupyter-troubleshoot.exe*
jupyter-trust.exe*
jupyter.exe*
```

**Comment by kalefranz** [comment]
CC @msarahan

**Comment by kalefranz** [comment]
relates to #2498

**Comment by quasiben** [comment]
If we did this I would want the preamble for `ipython` to include env information.  I've made this mistake a number of times (and probably will continue to do so) that a little note informing me would be super helpful.

For reference the current preamble looks like the following:

```
Python 2.7.11 |Continuum Analytics, Inc.| (default, Dec  6 2015, 18:08:32)
Type "copyright", "credits" or "license" for more information.
```

Suggest something like

```
Python 2.7.11 |Continuum Analytics, Inc.| (default, Dec  6 2015, 18:08:32)| ENV_INFORMATION
Type "copyright", "credits" or "license" for more information.
```

**Comment by kalefranz** [comment]
We'd have to hack/patch `sys.version` in the stdlib to do that.  But that's a good idea regardless of if we implement this issue or not.  Could probably be done.

**Comment by jakirkham** [comment]
I'm +0.5 on this change.

In some cases like OS X and Windows this is basically an old bug being turned into a feature. In some cases, `$PATH` is getting scrubbed more vigorously with `conda` and so this gives users an out so they can get back to old behavior (we were affected by this on Linux recently as we place `git` there 😉).

Though it is a little risky as others have pointed out because one doesn't really know where things are being loaded from. This could cause some surprising linking issues that will be fun to debug. Hence, why I'm a bit leery of this change. Particularly if it is done by default.

A long time ago there was an idea of having a `default` environment instead of `root`. This change feels similar to me in some ways. It is probably worth unearthing some of those issues. There is almost certainly something we can learn from [re]visiting those discussions like issue ( https://github.com/conda/conda/issues/1145 ).

My personal feeling would be to add a config setting to `.condarc` and default it to `false`. This way people have the option for getting `root` on the path. However, given the subtle problems that occur when this goes wrong, we should have them actually enable this option. Also, this allows for a nice issue resolution strategy ("Did you try disabling `root` from being on the path?").

Some other footnotes worth mentioning that probably need to find a home in this issue tracker. We had been discussing having some sort of executable only environment for use during builds. This would mainly be useful for bootstrapped packages (e.g. `cmake` on Windows, `go`, `gcc`, and other compilers). Similar effects could be achieved through this proposal. However, this idea of "stacked" envs would probably be cleaner. Also, subenvironments came up for discussion previously. This has similarities to the idea here and is of particular value when handling things like Jupyter and a bunch of kernels.

**Comment by msarahan** [comment]
This would be a big improvement, but we should always keep in mind that some people will want absolutely nothing added to their PATH.  Transient additions to PATH are fine, but no modifications to .bashrc, or to the Windows environment variables (outside of a prompt).  Details like this are what make this really fun!

**Comment by kalefranz** [comment]
I understand the concern about "shadowing" executables. I think this is a basic skill command line users are just going to have to learn though.  Get bitten by it once, and learn.  Learn to use `which -a` and `type -a`.  The reason is that this isn't just something that involves conda.  It's a more general pattern of how OSes and PATHs work.

Even if I use homebrew, for example, to install python.  I now have on path both /usr/local/bin/python and /usr/bin/python.  I have on path two pips.  One I must use sudo with, one I absolutely shouldn't.  The shadowing argument, to me, is a fundamental user education issue, and not one we should try to patch around with non-standard hackiness.

**Comment by jakirkham** [comment]
Sorry my comment was really long. Let me give the cliff notes on this change precisely. I would be +1 if there was a setting in `.condarc`/`conda config` to control this behavior and it was set to `false` by default.

**Comment by quasiben** [comment]
@kalefranz, I agree that it's a basic unix skill, but I don't know that I would hold windows users to the same standard.  For this, I agree with @jakirkham that a setting would be nice and it may potentially offer conda devs an easy way to debug/inform new conda and conda CLI users who may come here and declare a *bug*

**Comment by jakirkham** [comment]
> a setting would be nice and it may potentially offer conda devs an easy way to debug/inform new conda and conda CLI users who may come here and declare a bug

👍 Couldn't have said it better myself.

**Comment by njalerikson** [comment]
Given the rework of activate/deactivate done in #3175  the PATH only modifies what it has previously prepended to the PATH.

```
> echo $PATH
~/anaconda3/bin:/other/paths
>
> . activate ENVNAME
> echo $PATH
~/anaconda3/envs/ENVNAME/bin:~/anaconda3/bin:/other/paths
>
> . activate ENVNAME_2
> echo $PATH
~/anaconda3/envs/ENVNAME_2/bin:~/anaconda3/bin:/other/paths
>
> . deactivate
> echo $PATH
~/anaconda3/bin:/other/paths
>
> . activate root
> echo $PATH
~/anaconda3/bin:~/anaconda3/bin:/other/paths
>
> . deactivate
> echo $PATH
~/anaconda3/bin:/other/paths
```

**Comment by asmeurer** [comment]
This is what it used to do, and it was a mess. People would create environments that accidentally didn't include Python or some other package, and it would fallback to root, leading to much confusion. Even advanced users and yours truly did it on accident from time to time. Once this was changed, a huge chunk of support requests suddenly stopped happening. I highly recommend against this change. The real bug here is that Windows (still!?) has this incorrect behavior. 

And by the way, I don't understand why you want "stacked" ends. That's a bad abstraction from other package managers that conda's design obviates the need for, IMHO.

**Comment by msarahan** [comment]
The only reason windows still does this is that we need to keep the activate/deactivate scripts on PATH.  There is no other reason.  We have discussed ways to do that, but have not settled on a final solution yet.  I think that this will be solved soon, when conda moves into its own dedicated space - that space can remain on PATH, while keeping the other environments separate.

**Comment by njalerikson** [comment]
There appears to be two distinctive camps for this.

Maybe that should suggest a .condarc switch, keep_root_on_activate.

**Comment by asmeurer** [comment]
I forget, but does it clear the root env if it's on the PATH twice?  If not, then then anyone who really wanted this behavior (though I still don't see why) could just do that.

**Comment by njalerikson** [comment]
The way I have reworked the `activate`/`deactivate` logic means a few things:

1. The `deactivate` script only deactivates an environment if it finds that the environment variable CONDA_PREFIX exists (this is set in `activate` and likewise unset in `deactivate`). In other words, the root environment cannot be removed from the user's PATH unless they are explicitly doing something weird like setting CONDA_PREFIX (in which case I would think it's the user's fault for getting weird results).
2. The `activate` script always prepends the activating environment to the PATH. So in the case of `. activate root` a normal user will expect to see the root path(s) listed twice.
2. The `deactivate` script only removes the first occurrence of the environment's paths to deactivate. So again if the user activated root only the first occurrence of the root path will be removed.

So in conclusion, yes, if a user follows the recommended installation of Conda and appends the root environment to PATH they will be stacking the activating environment on top of the root environment. If they don't append the Conda root to their PATH then activating an environment will not stack it with the root. BUT consequently that also means that they cannot easily call the `activate`/`deactivate` scripts especially since these scripts (and their helper scripts) are currently designed to be available on the user's PATH.

**Comment by abretaud** [comment]
Hi,
I opened #3580 because we would need to stack not only the root env with one env, but also multiple environments.
I'm aware that in some cases it can lead to some incompatibilities between envs, but as cluster admins that would greatly ease the transition to conda.
I think having an option for this (in activate and/or in condarc?), just like the proposed keep_root_on_activate, even with a scary name and clear warnings in the doc, would be a good compromise.

**Comment by njalerikson** [comment]
After sleeping on this for awhile I've decided that the following plan will address all of the issues that I am aware of.

1. Remove symlinking. This is easily said, and mostly easily done. Need to consider existing users who have environments that already have symlinks created. Symlinks are currently recreated every time a user activates an environment and are left upon deactivation. So for a few Conda versions there would need to be a symlink removal clause.
2. Move all conda critical executables into a new "global" location. Something like `~/anaconda/global/` or `~/anaconda/sbin/`. This location will always be on the user's $PATH. I see no issue leaving this always on the `$PATH` since it will contain Python-agnostic code. 
2. Introduce a new environment variable/activate switch/.condarc value, `stack_envs`. This value will decide whether or not the `root` bin is left on the user's `$PATH` on activate.

The `stack_envs` variable will be perfect for phasing into a solution that will also work for #3580.

**Comment by asmeurer** [comment]
That doesn't sound so easy. If you remove the symlinks, then people won't have access to `activate`, `deactivate`, or `conda`, unless you create the separate conda environment first and put it on the PATH (permanently, i.e., you'd have to edit the bash profile like the installer does). 

And what would be the default value for `stack_envs`? False, I'd hope. The average use-case does not look like https://github.com/conda/conda/issues/3580.

**Comment by alexhsamuel** [comment]
I think this change is a bad idea. In response to the original issue, if I install git in my root and then activate another environment, then yes, I absolutely do think that git should no longer be in my `PATH`. I don't think the root environment should be special in any way.  Users who are confused by this, to quote @kalefranz, should learn to use `which` and `type`.

In particular, conda's attempt to seal up its environment interacts badly with the two stacked environments in `PATH`. Suppose I'm in my root environment, its `bin` is in my `$PATH` (as conda encourages), and,

```sh
$ python -c 'import affine'
ImportError: No module named 'affine'
$ conda create -n foo python affine
$ source activate foo
$ python -c 'import affine'  # Yay!
$ ipython -c 'import affine'
ImportError: No module named 'affine'  # Huh??
```

Same thing for `jupyter-notebook`, `py.test`, and a host of other things. The problem here is not that I am ignorant of how `PATH` works. The problem is that conda has rewritten the #! lines in the programs I'm running, explicitly to ignore my `PATH`! If the first line of `py.test` was `#!/usr/bin/env python`, things would be fine—I'd be testing with the `python` I explicitly indicated I wish to use (by activating its environment and thus putting it at the front of my `PATH`). But conda, in trying to prevent me from shooting myself in the foot, has made me shoot myself in the other foot.

As a side note, in the example above, even if I figure out what's going on it still may not work.

```sh
$ conda install ipython
$ ipython -c 'import affine'
ImportError: No module named 'affine'   # Whaaaaa!?
$ which ipython
~/conda/envs/foo/bin/ipython  # I'm so confused...
$ type ipython
ipython is hashed (~/conda/bin/ipython)
```

What fraction of users understand that `which` != `type` and they may need to `hash -r` before their conda install takes effect? Shadowing is complicated, shadowing + caching is truly evil.

It might be a cool feature for conda to support stacked environments, but I think a lot of thought would have to go into getting this right. In the meantime, I don't think it's right to give special treatment to the root environment in `PATH`.

I'd take @njalerikson's proposal one step further: I think it would be cleanest if conda wasn't self-hosting, _i.e._ if `conda` was a program that would live in `/usr/bin` or somewhere else, completely independent of any python executable or environment at all. That would remove the obstacle from any specialness of the root environment. I suspect this is a rather substantial change, however.

**Comment by asmeurer** [comment]
That's right. Shell hashing seems mostly unrelated here, but you are right that conda environments are designed to be isolated from one another. The real way to implement some kind of "stacking" in conda would be to reinstall each "lower" environment's packages, and provide some kind of interface on top of that to autoupdate envs. For instance, something like

```
env a: python, numpy
env b (stacked on top of env a): python, numpy, scipy

conda install -n a matplotlib  # installs matplotlib in both env a and env b
```

In other words, conda requires that env a's packages are always a subset of env b's packages when manipulating env a (and records some kind of "diff" when manipulating env b). I personally don't know if I would use the feature, but others seem to want it.

There's a lot more to how packages interact with one another than the PATH (even if conda packages did use `/usr/bin/env` in their shebang). There's also the Python path, dynamic library linking, <other language> paths, etc. You could make 90% of the cases work by setting 100 environment variables, but then you'd get a system that's 100x overengineered and still doesn't work for the remaining 10%.  

I'd reiterate: installing the same package into multiple environments with conda is extremely cheap (with the exception of cross filesystem installations and some special Linux filesystems that don't support hard linking).

**Comment by alexhsamuel** [comment]
How about `conda create --add ...` that creates a new environment with the same packages as the root, plus whatever else is specified?

**Comment by asmeurer** [comment]
I'm not a conda dev (any more), so I'll leave the implementation up to them (and it should probably be discussed at https://github.com/conda/conda/issues/3580).

**Comment by alanhdu** [comment]
FWIW (and this might be several days too late), I think @njalerikson's plan would be great (as long as stacking environments is off by default), but the root environment should just the environment that's activated by default, without any other special treatment.

Right now, when I want to use the system python (and the system shared objects) for sysadmin stuff, I have to manually edit my $PATH to leave out conda. It'd be great if I can just `source deactivate` (or maybe `source activate root && source deactivate`), do my sysadmin stuff, and then `source activate root` to get back to conda land.

**Comment by asmeurer** [comment]
@alanhdu I think you are looking for https://github.com/conda/conda/issues/813, which is related but distinct issue.

**Comment by kalefranz** [comment]
Closing this issue with behavior as settled for conda 4.6, which makes the `conda` command available everywhere, even if the base environment is not and/or never has been active (most work done in conda 4.4), along with not stacking activated environments by default, but adding a `--stack` flag to `conda activate` as discussed in https://github.com/conda/conda/issues/7194.
